### PR TITLE
executor: Add Global to SHOW INDEX output | tidb-test=pr/2391

### PR DESCRIPTION
**Pull request by samba-rgb**
### What problem does this PR solve?
No info about index if its global or not , added in show index in this pr

Issue Number: ref #55452

Problem Summary:

### What changed and how does it work?

added Global index from index info

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/show.go [hunk 1]
```diff
@@ -805,6 +807,14 @@ func (e *ShowExec) fetchShowIndex() error {
 		if colStats != nil {
 			ndv = colStats.NDV
 		}
+
+		pi := tb.Meta().GetPartitionInfo()
+		isPartitioned := pi != nil && pi.Type != pmodel.PartitionTypeNone
+
+		if isPartitioned && e.Ctx().GetSessionVars().EnableGlobalIndex {
+			isGlobalIndex = "YES"
```
Comment (mjonss) [inline pkg/executor/show.go:815]: If PKIsHandle, then it is by definition a LOCAL index, i.e. each partition will have the clustered index, and the PK column must be a part of the partitioning expression.

Test by:
```sql
set tidb_enable_global_index = 0;
create table t (a int primary key clustered, b int) partition by hash (a) partitions 2;
show create table t;
show index from t;
set tidb_enable_global_index = 1;
show index from t;
```

File: pkg/executor/show.go [hunk 2]
```diff
@@ -822,6 +832,7 @@ func (e *ShowExec) fetchShowIndex() error {
 			"YES",            // Index_visible
 			nil,              // Expression
 			"YES",            // Clustered
+			isGlobalIndex,    // Global_index
```
Comment (mjonss) [inline pkg/executor/show.go:835]: So here you can just use "NO"
```suggestion
			"NO",    // Global_index
```

File: pkg/executor/show_test.go [hunk 3]
```diff
@@ -97,3 +97,32 @@ func TestShow(t *testing.T) {
...
 }
+
+func TestShowIndex(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+
+	tk.MustExec("use test")
+	tk.MustExec("create table t(id int, abclmn int);")
+
+	defer tk.MustExec("drop table t;")
```
Comment (mjonss) [inline pkg/executor/show_test.go:108]: No need to remove tables, each MockStore will be created fresh
```suggestion 
```

**Review by mjonss — COMMENTED** [review]
Looks good, with some comments.
There are also more tests that needs to update (since the SHOW INDEX output changed).
You can run all unit tests with `make dev` and you can run integration tests by first building the `tidb-server` binary with `make` and then `(cd tests/integrationtest ; ./run-tests.sh -s ../../bin/tidb-server)`

**Comment by mjonss** [comment]
@samba-rgb I can fix the idc-jenkins-ci-tidb/mysql-test test failures. Can you fix the other failed tests?

**Comment by samba-rgb** [comment]
@mjons Thank you for helping with my SQL test fix. I will now resume working on the fixes for integration and unit tests, along with addressing the comments mentioned above.

File: pkg/executor/show.go (hunk 2 above)
Comment (mjonss) [reply pkg/executor/show.go]: Ah, sorry, I had not indented it correctly :(

**Comment by samba-rgb** [comment]
Hi @mjonss i added the  Global column in the integrationtest/r/*/*.result path
 example : tidb/tests/integrationtest/r/session/clustered_index.result
 but still in the check_dev its not reflecting 
 is their any place that i need to change/add
 
 BTW when i ran make dev no test case is running for me 
 find: md5sum: No such file or directory

**Comment by mjonss** [comment]
> Hi @mjonss i added the Global column in the integrationtest/r/_/_.result path example : tidb/tests/integrationtest/r/session/clustered_index.result but still in the check_dev its not reflecting is their any place that i need to change/add
> 
> BTW when i ran` make dev` no test case is running for me find: md5sum: No such file or directory

@samba-rgb 
For the integration tests, you can simply re-record the result files, it seems like there is some formatting issues in your change, likely spaces instead of tabs. Try `./run-tests.sh -s ../../bin/tidb-server -r session/clustered_index` and it will overwrite the .result file, and you can try with `./run-tests.sh -s ../../bin/tidb-server -t session/clustered_index` to verify that it passes. 

I assume you are using a mac? md5sum is md5 in macOs..., you can also run only the unit tests with `make ut` IIRC.
Also if there are a few integration tests that fails on mac, it is due to floating point difference between arm and i386, so leave them as-is.

I will try to find which unit tests still needing to be updated.

**Comment by mjonss** [comment]
pkg/ddl/db_change_test.go:857
        	Error:      	Received unexpected error:
        	            	need [[t 0 PRIMARY 1 c1 A 0 <nil> <nil>  BTREE   YES <nil> NO]], but got [[t 0 PRIMARY 1 c1 A 0 <nil> <nil>  BTREE   YES <nil> NO NO]]
        	Test:       	TestShowIndex

To run the test you must also enable failpoints, `make failpoint-enable`, which you then can disable before committing, `make failpoint-disable`
Maybe this patch file could help?
[56028.patch](https://github.com/user-attachments/files/17050432/56028.patch)

**Review by lance6716 — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**